mirgecom/flux.py

```diff
@@ -30,6 +31,59 @@
 THE SOFTWARE.
 """
 import numpy as np  # noqa
+from meshmode.dof_array import DOFArray
+from mirgecom.fluid import (
+    ConservedVars,
+    make_conserved
+)
+
+
+def central_scalar_flux(trace_pair, normal):
```

Comment:
This function calculates the central flux for a vector of scalars, so its name is appropriate.  The fact that we use it when calculating the gradient of a vector of scalars is independent of this actual routine and what it does.  We could just as well use a different numerical flux for calculating the grad.  In general I do not agree with the notion that we should name routines after what they are used for - but instead about what they *do* or what they *are*.

The numerical fluxes in H+W are defined in terms of the `average` and `jump` operators, which are explained for both vector and scalar quantities here:
<img width="523" alt="Screen Shot 2021-08-19 at 1 31 17 PM" src="https://user-images.githubusercontent.com/8116930/130124709-1d7cfbd4-e6d5-410d-9ce6-5bbe492af7e2.png">
   
This is essentially what drives my choice of whether I name the thing a "scalar" flux or a "vector" flux.

Reply:
> The fact that we use it when calculating the gradient of a vector of scalars is independent of this actual routine and what it does. We could just as well use a different numerical flux for calculating the grad.

I don't think this is true. The `u*normal` form of the flux pops out as the thing getting multiplied by the test function inside the surface integral when deriving the DG scheme for `s = grad(u)`.

> In general I do not agree with the notion that we should name routines after what they are used for - but instead about what they _do_ or what they _are_.

I don't know how to make sense of a flux without reference to some operator/equation/physics...

> The numerical fluxes in H+W are defined in terms of the `average` and `jump` operators, which are explained for both vector and scalar quantities here:
> <img alt="Screen Shot 2021-08-19 at 1 31 17 PM" width="523" src="https://user-images.githubusercontent.com/8116930/130124709-1d7cfbd4-e6d5-410d-9ce6-5bbe492af7e2.png">
> 
> This is essentially what drives my choice of whether I name the thing a "scalar" flux or a "vector" flux.

The numerical fluxes are defined by taking the term in the surface integral and modifying it with averages, jumps, etc. in order to couple the elements together.

Reply:
I think I see the disconnect here. I'm loosely calling anything that appears as a surface integral over the element (i.e. as a result of divergence theorem, integration by parts, or similar action) as a "flux". Such flux terms can be constructed for arbitrary fields as things in-and-of-themselves at the element boundary - and are different from the transport fluxes.

Additionally, there is the reconstruction method for the quantities that we intend to integrate over the element surface, which are needed because the fields are not (in-general) single-valued at the element faces.  I am (currently) naming these routines according to the reconstruction method and the input field - not the intended use of the constructed quantity.  Describing fluxes according to the reconstruction method is a common practice in the literature, as well. 

We have 3 different numerical flux routines used in the fluid code right now.  

- One that implements what BR1 calls *H_s* in their eqn 13 - for the aux eqn.  I'm OK with calling that `grad_flux` if you want, but to me that name is less descriptive of what the routine is doing. They prescribe a central reconstruction of a scalar field there (i.e. treating each component of *u* as an independent scalar field).
- One that implements BR1's *h_e* in their eqn 10, which is the flux term resulting from applying the divergence theorem/IBP to the volume integral of the divergence of the inviscid part of the transport flux.  They suggest several, we use a LLF numerical flux there.
- One that implements BR1's *h_v* in their eqn 15, which is the flux term resulting from applying the divergence theorem/IBP to the volume integral of the divergence of the viscous part of the transport flux.  They prescribe a central reconstruction for the viscous flux vector, and then dot that with the face normals to form the "flux" that is to be integrated over the element surface.

If we rename, for example, the `central_vector_flux` to `divergence_flux`, then what to do with the LLF routine?  It too is a flux intended for use in a divergence integral. This change would suggest we should call both of these the `divergence_flux` and then pass in the reconstruction method as either central, or LLF, or other.  But this will be complicated by the presence of penalty terms in some reconstructions and not others.     

I am not opposed to reconstructing things like that - but I do suggest that we should take that flux refactoring on in a dedicated effort.

Reply:
> I think I see the disconnect here. I'm loosely calling anything that appears as a surface integral over the element (i.e. as a result of divergence theorem, integration by parts, or similar action) as a "flux".

Me too. 🙂 

> Such flux terms can be constructed for arbitrary fields as things in-and-of-themselves at the element boundary - and are different from the transport fluxes.

Sure, but they're still induced by the operator being discretized, so I think it makes sense to tie them to the `grad`/`div` operators. Off the top of my head I can't picture a way that these particular flux forms would arise otherwise.

> Additionally, there is the reconstruction method for the quantities that we intend to integrate over the element surface, which are needed because the fields are not (in-general) single-valued at the element faces. I am (currently) naming these routines according to the reconstruction method and the input field - not the intended use of the constructed quantity. Describing fluxes according to the reconstruction method is a common practice in the literature, as well.
> 
> We have 3 different numerical flux routines used in the fluid code right now.
> 
> * One that implements what BR1 calls _H_s_ in their eqn 13 - for the aux eqn.  I'm OK with calling that `grad_flux` if you want, but to me that name is less descriptive of what the routine is doing. They prescribe a central reconstruction of a scalar field there (i.e. treating each component of _u_ as an independent scalar field).
> * One that implements BR1's _h_e_ in their eqn 10, which is the flux term resulting from applying the divergence theorem/IBP to the volume integral of the divergence of the inviscid part of the transport flux.  They suggest several, we use a LLF numerical flux there.
> * One that implements BR1's _h_v_ in their eqn 15, which is the flux term resulting from applying the divergence theorem/IBP to the volume integral of the divergence of the viscous part of the transport flux.  They prescribe a central reconstruction for the viscous flux vector, and then dot that with the face normals to form the "flux" that is to be integrated over the element surface.

Note that all of the BR1 names refer to the operator to which they correspond. I'm fine with `central_grad_flux` if that's your objection (it's unclear to me right now whether we could make a general `grad_flux` function that takes a `central_flux` function (say) as a parameter, or if maybe it would work the other way around; we only have one gradient flux at the moment, though, so I'm not really giving it a lot of thought). It's the `scalar`/`vector` terminology that I don't like; it doesn't tell me anything about what the flux represents.

> If we rename, for example, the `central_vector_flux` to `divergence_flux`, then what to do with the LLF routine? It too is a flux intended for use in a divergence integral. This change would suggest we should call both of these the `divergence_flux` and then pass in the reconstruction method as either central, or LLF, or other. But this will be complicated by the presence of penalty terms in some reconstructions and not others.
> 
> I am not opposed to reconstructing things like that - but I do suggest that we should take that flux refactoring on in a dedicated effort.

I just noticed that `lfr_flux` in the code is something different from what I thought. In the book it's just:

![Screen Shot 2021-08-20 at 4 00 11 PM](https://user-images.githubusercontent.com/7232918/130292842-75b607f6-4555-4c46-8712-834826c61305.png)

which is (in principle at least) independent of the operator. But the current implementation explicitly encodes the flux for a divergence. I would advocate for either:

1. Renaming `lfr_flux` to something like `lfr_div_flux` and putting it in `operators.py` (along with renaming `central_vector_flux` to `central_div_flux` and sticking it in there too).
2. Passing `f` in as a function parameter, and leaving it as `lfr_flux` in `flux.py`.

(But this is probably a job for a separate PR. Let's limit the scope to `central_scalar_flux` here.)

Reply:
> > Such flux terms can be constructed for arbitrary fields as things in-and-of-themselves at the element boundary - and are different from the transport fluxes.
> 
> Sure, but they're still induced by the operator being discretized, so I think it makes sense to tie them to the `grad`/`div` operators. Off the top of my head I can't picture a way that these particular flux forms would arise otherwise.

Fair enough. I agree that it does make sense to tie them to the operators, and I am happy to rename them as such.

> Note that all of the BR1 names refer to the operator to which they correspond. 

I don't see that. Can you point it out?

> I'm fine with `central_grad_flux` if that's your objection (it's unclear to me right now whether we could make a general `grad_flux` function that takes a `central_flux` function (say) as a parameter, or if maybe it would work the other way around; we only have one gradient flux at the moment, though, so I'm not really giving it a lot of thought). 

I guess I was thinking we could have a `gradient_flux` and then pass in the reconstruction method such as: 

```python
def central_reconstruction(u_tpair):
    return u_tpair.avg
```

Clearly, the central reconstruction is trivial esp with the TracePair support, but more complicated reconstruction routines can easily be imagined. In the case of transport flux reconstructions there are clearly many varied choices.   

> It's the `scalar`/`vector` terminology that I don't like; it doesn't tell me anything about what the flux represents.

Absolutely true of course. Again, the name was designed to say what it was doing, not what it would be used for.

What I ended up doing was renaming both - since this is the easiest point at which to do so. We use them more extensively in the soon-to-follow-on developments. I hope they are making more sense from your perspective now.   (2c2c6e0)

Reply:
> > Note that all of the BR1 names refer to the operator to which they correspond.
> 
> I don't see that. Can you point it out?

The subscripts: `_s` for S (aka `grad(u)`), `_e` for Euler, `_v` for viscous.

> What I ended up doing was renaming both - since this is the easiest point at which to do so. We use them more extensively in the soon-to-follow-on developments. I hope they are making more sense from your perspective now. ([2c2c6e0](https://github.com/illinois-ceesd/mirgecom/commit/2c2c6e0393474d71c34de8efe18f334a2e184b02))

I think this looks pretty good for now. Eventually I think it would be good to have these in the same module as `div_operator`/`grad_operator` and name them similar to the other numerical fluxes (i.e., call them `div_facial_flux`/`grad_facial_flux`). And then we would just put the operator-independent stuff (like the generalized LFR function and `central_reconstruction`, etc.) in `flux.py`. But we can work on that later.